packages/storage-plus/src/stack.rs

```diff
@@ -0,0 +1,652 @@
+//! An "Stack" is a storage wrapper that guarantees constant-cost appending to and popping
+//! from a list of items in storage.
+//!
+//! This is achieved by storing each item in a separate storage entry. A special key is reserved
+//! for storing the length of the collection so far.
+use std::any::type_name;
+use std::convert::TryInto;
+use std::marker::PhantomData;
+use std::sync::Mutex;
+
+use serde::{de::DeserializeOwned, Serialize};
+
+use cosmwasm_std::{to_vec, StdError, StdResult, Storage};
+
+use crate::helpers::must_deserialize;
+
+const LEN_KEY: &[u8] = b"len";
+
+pub struct Stack<'a, T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    /// prefix of the newly constructed Storage
+    namespace: &'a [u8],
+    /// needed if any suffixes were added to the original namespace.
+    /// therefore it is not necessarily same as the namespace.
+    prefix: Option<Vec<u8>>,
+    length: Mutex<Option<u32>>,
+    item_type: PhantomData<T>,
+}
+
+impl<'a, T: Serialize + DeserializeOwned> Stack<'a, T> {
+    /// constructor
+    pub const fn new(prefix: &'a str) -> Self {
+        Self {
+            namespace: prefix.as_bytes(),
+            prefix: None,
+            length: Mutex::new(None),
```

Comment:
What do you need that mutex for? Should not be needed because CosmWasm is a single threaded environment